spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,492 @@
+)
+dl_valid = create_dict_dataloader(
+    X_valid, Y_valid, split="valid", batch_size=BATCH_SIZE, shuffle=False
+)
+dl_test = create_dict_dataloader(
+    X_test, Y_test, split="test", batch_size=BATCH_SIZE, shuffle=False
+)
+
+# %% [markdown]
+# We can inspect our datasets to confirm that they have the appropriate fields.
+
+# %%
+dl_valid.dataset
+
+# %% [markdown]
+# ### Define [`MultitaskClassifier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html)
+#
+# We define a simple Multi-Layer Perceptron (MLP) architecture to learn from the `bow_features`.
+#
+# _Note: the following might feel like extra steps to define what is a very simple architecture (e.g. `sklearn`'s [`MLPClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html)), but this will lend us additional flexibility later in the pipeline!_
+
+# %% [markdown]
+# To start, we define a `module_pool` with all the PyTorch modules that we'll want to include in our network.
+
+# %%
+import torch.nn as nn
+
+bow_dim = X_train.shape[1]
+module_pool = nn.ModuleDict(
+    {
+        "mlp": nn.Sequential(nn.Linear(bow_dim, bow_dim), nn.ReLU()),
+        "prediction_head": nn.Linear(bow_dim, 2),
+    }
+)
+
+# %% [markdown]
+# Then, we specify the desired `task_flow` through each module.
+
+# %%
+from snorkel.classification.task import Operation
+
+task_flow = [
+    Operation(name="input_op", module_name="mlp", inputs=[("_input_", "bow_features")]),
+    Operation(name="head_op", module_name="prediction_head", inputs=[("input_op", 0)]),
+]
+
+# %% [markdown]
+# With these pieces, we're ready to define a [`Task`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.html#module-snorkel.classification.task) in Snorkel for spam classification.
+
+# %%
+from functools import partial
+from snorkel.analysis import Scorer
+from snorkel.classification import (
+    Task,
+    cross_entropy_from_outputs,
+    softmax_from_outputs,
+)
+
+spam_task = Task(
+    name="spam_task",
+    module_pool=module_pool,
+    task_flow=task_flow,
+    loss_func=partial(cross_entropy_from_outputs, "head_op"),
+    output_func=partial(softmax_from_outputs, "head_op"),
+    scorer=Scorer(metrics=["accuracy", "f1"]),
+)
+
+# %% [markdown]
+# We'll initialize a `MultitaskClassifier` with the `spam_task` we've created, initialize a corresponding [`Trainer`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.Trainer.html), and `fit` to our dataloaders!
+
+# %%
+from snorkel.classification import MultitaskClassifier, Trainer
+
+model = MultitaskClassifier([spam_task])
+trainer = Trainer(n_epochs=5, lr=1e-4, progress_bar=True)
+trainer.fit(model, [dl_train, dl_valid])
+
+# %% [markdown]
+# How well does our model do?
+
+# %%
+model.score([dl_train, dl_valid], as_dataframe=True)
+
+# %% [markdown]
+# ## 3. Perform error analysis
```

Comment:
Do we need this now?

Reply:
I like this as an explicit step; emphasizes the workflow component

Reply:
ok well the text needs to change a bit...kind of dangling here. at least need a conclusion and how that flows into the monitoring section

Reply:
Added this:

> In the next sections, we'll explore how we can programmatically monitor these error modes with built-in helpers from Snorkel.